src/items.rs

```diff
@@ -113,17 +113,57 @@ impl Rewrite for ast::Local {
         result.push_str(&infix);
 
         if let Some((init, _els)) = self.kind.init_else_opt() {
+            let base_span = if let Some(ref ty) = self.ty {
+                mk_sp(ty.span.hi(), self.span.hi())
+            } else {
+                mk_sp(self.pat.span.hi(), self.span.hi())
+            };
+
+            let offset = context.snippet(base_span).find_uncommented("=")?;
+            let base_span_lo = base_span.lo();
+
+            let assign_lo = base_span_lo + BytePos(offset as u32);
+            let comment_start_pos = if let Some(ref ty) = self.ty {
+                ty.span.hi()
+            } else {
+                self.pat.span.hi()
+            };
+            let comment_before_assign = context.snippet(mk_sp(comment_start_pos, assign_lo)).trim();
+
+            let assign_hi = base_span_lo + BytePos((offset + 1) as u32);
+            let rhs_span_lo = init.span.lo();
+            let comment_end_pos = if init.attrs.is_empty() {
+                rhs_span_lo
+            } else {
+                let attr_span_lo = init.attrs.first().unwrap().span.lo();
+                // for the case using block
+                // ex. let x = { #![my_attr]do_something(); }
+                if rhs_span_lo < attr_span_lo {
+                    rhs_span_lo
+                } else {
+                    attr_span_lo
+                }
+            };
+
+            if !comment_before_assign.is_empty() {
+                let new_indent_str = &pat_shape
+                    .block_indent(0)
+                    .to_string_with_newline(context.config);
+                result = format!("{}{}{}", comment_before_assign, new_indent_str, result);
+            }
+
```

Comment:
Took some time to look into this code. It seems like a lot of what's going on here is to recover comments before the `=`, e.g. `let pat: ty /*comment*/ =` or `let pat /*comment*/ =`. However, this PR and all the test cases you've outlined below only test for comments that come after the `=` e.g. `let pat: ty = /*comment*/ expr`.

For that reason I think we should only worry about the post `=` comments in this PR.

I also propose renaming `init` to `expr`  since I think it's a little clearer that it's an `ast::Expr` node, and using `init()` instead of `init_else_opt()` since we don't use the `_else` and currently [don't support `ast::LocalKind::InitElse(..)` variants](https://github.com/rust-lang/rustfmt/blob/ef91154250977b3b5d05448dafbca524a1168b47/src/items.rs#L57-L59).

```rust
if let Some(expr) = self.kind.init() {
    // Adjust the span after the `pat` and `ty` before searching for the `=`
    let base_span = if let Some(ref ty) = self.ty {
        mk_sp(ty.span.hi(), self.span.hi())
    } else {
        mk_sp(self.pat.span.hi(), self.span.hi())
    };

    let comment_lo = context.snippet_provider.span_after(base_span, "=");
    let comment_span = mk_sp(comment_lo, expr.span.lo());

    let nested_shape = shape.sub_width(1)?; // I don't think this is necessary
    result = rewrite_assign_rhs_with_comments(
        context,
        result,
        expr,
        nested_shape,
        &RhsAssignKind::Expr(&expr.kind, expr.span),
        RhsTactics::Default,
        comment_span,
        true,
    )?;
}
```

Reply:
> ... a lot of what's going on here is to recover comments before the `=`

I somehow missed that when I copied the R2 commit changes :frowning_face:.  I also checked now R2 and didn't find any test to the assign pre `=` comments.  I therefore assume that maybe that commit was tested properly.

Changed the code per suggestion.